### PR TITLE
Documentation Sprint

### PR DESCRIPTION
**Pull request by cwfitzgerald**
In an attempt to force all new code to have everything fully documented, we must first document everything. This does exactly that. This adds a basic level of documentation to everything. I suggest you mainly look for factual inaccuracies or blatant spots where documentation could be trivially improved.

This will now fail PRs when they merge if they don't have docs, or will issue warnings until you add documentation.

### CODE REVIEW

File: src/shaderpack/mod.rs [hunk 1]
```diff
@@ -1,91 +1,125 @@
...
 use std::path::PathBuf;
 
+/// A fully parsed Nova Shaderpack
 #[derive(Debug, Clone)]
 pub struct ShaderpackData {
+    /// Pipelines described by the shaderpack.
     pipelines: Vec<PipelineCreationInfo>,
-    /// All the renderpasses that this shaderpack needs, in submission order
+
+    /// All the renderpasses that this shaderpack needs, in submission order.
```
Comment (DethRaid) [inline src/shaderpack/mod.rs]: I wouldn't say that they're the renderpasses that the shaderpack _needs_, but that they're the specification for needed renderpasses

File: src/shaderpack/mod.rs [hunk 2]
```diff
@@ -391,6 +494,11 @@ impl TextureFormat {
         0.0
     }
 
+    /// Returns the screen size in pixels.
```
Comment (DethRaid) [inline src/shaderpack/mod.rs]: Returns the texture size in pixels

File: src/shaderpack/mod.rs [hunk 3]
```diff
@@ -401,184 +509,297 @@ impl TextureFormat {
     }
 }
 
+/// State of the fixed-function rasterizer.
```
Comment (DethRaid) [inline src/shaderpack/mod.rs]: Not exactly - this is an enum of possible rasterizer hardware features that a shaderpack can enable or disable

It's a weird Bedrock thing and we should move this to a bunch of booleans or something on the Pipeline struct

File: src/shaderpack/mod.rs [hunk 4]
```diff
@@ -401,184 +509,297 @@ impl TextureFormat {
...
     ///
-    /// 8 bytes (might try 4)
+    /// 8 bytes (might try 4).
     UV0,
-    ///  The UV coordinate in the lightmap texture
+
+    /// The UV coordinate in the lightmap texture.
     ///
-    /// This is a real UV and it doesn't change for no good reason
+    /// This is a real UV and it doesn't change for no good reason.
```
Comment (DethRaid) [inline src/shaderpack/mod.rs]: I wish I remember why I wrote this line lol

File: src/rhi/rhi_enums.rs [hunk 5]
```diff
@@ -1,251 +1,467 @@
...
     FragmentShaderReadOnly,
 
+    /// The source of a transfer.
     TransferSource,
+
+    /// The destination of a transfer.
     TransferDestination,
 }
 
+/// Type of buffer current descriptor points to.
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply src/rhi/rhi_enums.rs]: What should I use to describe it?

File: src/rhi/rhi_traits.rs [hunk 6]
```diff
@@ -15,233 +15,268 @@ use crate::surface::Surface;
...
         &self,
         pipeline_interface: Self::PipelineInterface,
         data: shaderpack::PipelineCreationInfo,
     ) -> Result<Self::Pipeline, PipelineCreationError>;
 
-    /// Creates an Image from the specified ImageCreateInto
+    /// Creates an Image from the specified ImageCreateInto.
     ///
-    /// Images are created directly from the Device and not from a MemoryPool because
+    /// FIXME(dethraid): Is this true anymore? If not does this need to change the structure
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply src/rhi/rhi_traits.rs]: Do you want this to be part of this PR?

File: src/shaderpack/mod.rs (hunk 2 above)
Comment (cwfitzgerald) [reply src/shaderpack/mod.rs]: Typo

File: src/shaderpack/mod.rs (hunk 4 above)
Comment (cwfitzgerald) [reply src/shaderpack/mod.rs]: Lmao

File: src/rhi/rhi_enums.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (DethRaid) [reply src/rhi/rhi_enums.rs]: It's when a descriptor refers to both an image and a sampler

https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#descriptorsets-combinedimagesampler talks about it

File: src/rhi/rhi_traits.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply src/rhi/rhi_traits.rs]: Ah, I love the sound of "that's someone else's problem" lol

File: src/rhi/rhi_structs.rs [hunk 7]
```diff
@@ -2,104 +2,149 @@ use super::{rhi_enums::*, rhi_traits::*};
...
+    /// The resource the barrier is guarding.
     pub resource: Arc<dyn Resource>,
 
+    /// Initial resource state before the barrier.
     pub initial_state: ResourceState,
 
+    /// Final resource state after the barrier.
     pub final_state: ResourceState,
 
+    /// Accessibility before barrier.
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply src/rhi/rhi_structs.rs]: Leaving it for now, once we figure out something better we can amend it.

File: src/shaderpack/mod.rs (hunk 3 above)
Comment (cwfitzgerald) [reply src/shaderpack/mod.rs]: I feel like this should also be a bitflag, but that's out of scope for this PR.

File: src/shaderpack/mod.rs [hunk 8]
```diff
@@ -11,7 +11,7 @@ pub struct ShaderpackData {
     /// Pipelines described by the shaderpack.
     pipelines: Vec<PipelineCreationInfo>,
 
-    /// All the renderpasses that this shaderpack needs, in submission order.
+    /// All the renderpasses that this shaderpack specifies, in submission order.
```
Comment (DethRaid) [inline src/shaderpack/mod.rs]: Could you add this language to the other fields in this struct?

File: src/rhi/rhi_enums.rs [hunk 9]
```diff
@@ -204,7 +204,8 @@ pub enum ResourceState {
     /// The resource may be used for anything you want, but it won't be optimal for anything.
     General,
 
-    /// Attached to the color attachment of a texture.
+    /// Attached to the color attachment of a texture. Optimized to be a framebuffer
```
Comment (DethRaid) [inline src/rhi/rhi_enums.rs]: This `ResourceState` isn't that a resource _is_ attached to something, it's that the resource is optimized for being a color attachment. Could you remove the "Attached to the color attachment of a texture", and change the docs for other enum constants to also have this change?

File: src/rhi/rhi_enums.rs [hunk 10]
```diff
@@ -204,35 +205,34 @@ pub enum ResourceState {
     /// The resource may be used for anything you want, but it won't be optimal for anything.
     General,
 
-    /// Attached to the color attachment of a texture. Optimized to be a framebuffer
-    /// color attachment.
+    /// Resource optimized to be a framebuffer color attachment.
     ColorAttachment,
 
-    /// Attached to the depth-stencil attachment of a texture.
+    /// Resource optimized to be a depth-stencil attachment of a texture.
```
Comment (DethRaid) [inline src/rhi/rhi_enums.rs]: depth-stencil attachment _of a framebuffer_

Comment (DethRaid) [reply src/rhi/rhi_enums.rs]: Textures don't _have_ attachments, they _are_ attachments

File: src/shaderpack/mod.rs [hunk 11]
```diff
@@ -8,16 +8,16 @@ use std::path::PathBuf;
 /// A fully parsed Nova Shaderpack
 #[derive(Debug, Clone)]
 pub struct ShaderpackData {
-    /// Pipelines described by the shaderpack.
+    /// The pipeline that this shaderpack specifies.
```
Comment (DethRaid) [inline src/shaderpack/mod.rs]: The pipeline_s_ that this shaderpack specifies

**Pull request merged**